tracking/translations_parser/publishers.py

```diff
@@ -0,0 +1,137 @@
+import csv
+import logging
+from abc import ABC
+from pathlib import Path
+from typing import Sequence
+
+import wandb
+from translations_parser.data import Metric, TrainingEpoch, TrainingLog, ValidationEpoch
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)s] %(message)s",
+)
+logger = logging.getLogger(__name__)
+
+
+class Publisher(ABC):
+    """
+    Generic mixin to publish parsed data.
+    Either the `handle_*` methods can be overriden for real time publication or
+    the `publish` method with all results (including run date, configuration…).
+    """
+
+    def open(self, parser) -> None:
+        ...
+
+    def handle_training(self, training: TrainingEpoch, **kwargs) -> None:
+        ...
+
+    def handle_validation(self, validation: ValidationEpoch) -> None:
+        ...
+
+    def handle_metrics(self, metrics: Sequence[Metric]) -> None:
+        ...
+
+    def publish(self, log: TrainingLog) -> None:
+        ...
+
+    def close(self) -> None:
+        ...
+
+
+class CSVExport(Publisher):
+    def __init__(self, output_dir):
+        assert output_dir.is_dir(), "Output must be a valid directory"
+        self.output_dir = output_dir
+
+    def write_data(self, output, entries, dataclass):
+        if not entries:
+            logger.warning(f"No {dataclass.__name__} entry, skipping.")
+        with open(output, "w") as f:
+            writer = csv.DictWriter(f, fieldnames=dataclass.__annotations__)
+            writer.writeheader()
+            for entry in entries:
+                writer.writerow(vars(entry))
+
+    def publish(self, training_log):
+        training_output = self.output_dir / "training.csv"
+        if training_output.exists():
+            logger.warning(f"Training output file {training_output} exists, skipping.")
+        else:
+            self.write_data(training_output, training_log.training, TrainingEpoch)
+
+        validation_output = self.output_dir / "validation.csv"
+        if validation_output.exists():
+            logger.warning(f"Validation output file {validation_output} exists, skipping.")
+        else:
+            self.write_data(validation_output, training_log.validation, ValidationEpoch)
+
+
+class WandB(Publisher):
```

Comment:
I agree on this, I did it in a generic way when implementing the CSV export initially. If we keep the API it should use real time publication only (an the inner logic can use close e.g. to write a file), but it could certainly be better. It would also be easier to work on this once this huge PR is merged.